subprojects/docs/src/docs/userguide/developing-plugins/writing_in_kotlin.adoc

```diff
@@ -0,0 +1,329 @@
+* configures the `kotlin-assignment` Kotlin compiler plugin
+* sets `apiVersion` and `languageVersion` to the Kotlin embedded version
+* adds the `-Xsam-conversions=class` compiler option
+* adds others compiler options for compatibility:
+** `-java-parameters` to support https://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Parameter.html[Java 8 Parameter] class and getting method parameters through reflection
+** `-Xjvm-default=all` to add link:https://kotlinlang.org/docs/java-to-kotlin-interop.html#default-methods-in-interfaces[Default methods in interfaces]
+** `-Xjsr305=strict` for https://kotlinlang.org/docs/java-interop.html#compiler-configuration[increased null safety]
+
+
+The `kotlin-dsl` plugin is the-all-in-one plugin. By applying the `kotlin-dsl-precompiled-script-plugins` and the same options as Gradle when compiling scripts, it allows you to write plugin like you would write scripts, using `.gradle.kts` file.
+
+Like the `embedded-kotlin` plugin, it uses the same Kotlin version as your Gradle build. This is useful if you do not intend to distribute your plugins, for an example, for convention plugins. If you need to distribute your plugin, make sure to use a Kotlin version that is compatible with the Gradle version you are targetting
+
+
+[[sec:your_public_api]]
+== Plugin API compatibility
+
+In order to make your plugin API compatible with Groovy, avoid relying on Kotlin specific features so that the developer experience is as good for Groovy scripts as for Kotlin scripts. For an example the Kotlin compiler emits https://github.com/JetBrains/kotlin/blob/master/libraries/kotlinx-metadata/jvm/ReadMe.md[metadata] that only the Kotlin compiler can understand for features like default parameters or extension function. It is OK to use those to provide a better experience for Kotlin scripts as long as same functionality is accessible through other means in Groovy.
+
+In general, relying on link:{kotlin-reference}java-interop.html[calling Java from Kotlin recommendations] and link:{kotlin-reference}java-to-kotlin-interop.html[calling Kotlin from Java recommendations] works well for Groovy too.
+
+[[sec:action]]
+`Action`
+
+Do not use Kotlin function types in your API. Groovy doesn't know about Kotlin anything about. Groovy uses Closures.
+
+Instead, use an `Action<T>` parameter. Decorated Gradle objects contain an overload for every function that takes an `Action<T>` parameter:
+
+[source,kotlin]
+----
+class Something
+
+abstract class MyExtension {
+    // Use Action<T> in your APIs
+    // Gradle will generate an overload at runtime that takes a Groovy Closure
+    fun doSomething(action: Action<Something>) {
+        //
+    }
+}
+----
+
+In order for the overload to be generated, the object must be constructed by Gradle. For an example:
+
+[source,kotlin]
+----
+abstract class MyPlugin: Plugin<Project> {
+  override fun apply(project: Project) {
+    project.extensions.create("myExtension", MyExtension::class.java)
```

Comment:
With the DSL dependency added, I guess this looks a bit nicer


```
    project.extensions.create<MyExtension>("myExtension")
```

Reply:
Given this is in the "public API" section, I went for the most compatible option. It will work no matter what plugins you use.